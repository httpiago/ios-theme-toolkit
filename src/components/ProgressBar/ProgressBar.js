import React from 'react'
import PropTypes from 'prop-types'
import { mapClassList } from '../../Utils'

/**
 * ProgressBar component.
 *
 * @since 0.1.0
 * @see https://httpiago.github.io/ios-theme-toolkit/#/components/ProgressBar
 */
export default function ProgressBar({ value, height, round, color, infinite, className: aditionalClasses, ...rest }) {

  const prefix = 'progressbar'
  const classes = mapClassList({
    [`${prefix}--${color}`]: color,
    [`${prefix}--round`]: round,
    [`${prefix}--infinite`]: infinite
  })
  let percentage = Math.max(0, Math.min(value, 100))

  return (
    <div
      className={`${prefix} ${classes} ${aditionalClasses}`}
      style={{ height }}
      {...rest}
    >
      <span style={{ width: percentage + '%' }} />
    </div>
  )
}

ProgressBar.defaultProps = {
  height: 3,
  round: false,
  color: 'blue',
  infinite: false,
  className: ''
}

ProgressBar.propTypes = {
  /** Percentage of progress. (0 - 100) */
  value: PropTypes.number.isRequired,
  /** Height of element */
  height: PropTypes.number,
  /** Round edges. */
  round: PropTypes.bool,
  /** Color of progress bar. */
  color: PropTypes.oneOf([ 'default', 'blue', 'red', 'green', 'yellow', 'gray', 'white' ]),
  /** Show an animation when load it is not forecast to end. */
  infinite: PropTypes.bool,
  /** Aditional classes. */
  className: PropTypes.string,
}
