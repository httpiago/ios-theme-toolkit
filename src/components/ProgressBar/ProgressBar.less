@import '../../customizations.less';

// Default style
.progressbar {
  background-color: #d6d6d6;
  position: relative;
  overflow: hidden;
  width: 100%;
  height: 3px;
}

.progressbar > span {
  background-color: @green-color;
  position: absolute;
  height: 100%;
  width: 0%;
  top: 0;
  left: 0
}

// Barra de progresso com cantos arredondados
.progressbar--round {
  border-radius: 50px;
}

// Barra de progresso azul
.progressbar--blue > span {
  background-color: @blue-color;
}

// Barra de progresso vermelha
.progressbar--red > span {
  background-color: @red-color;
}

// Barra de progresso verde
.progressbar--green > span {
  background-color: @green-color;
}

// Barra de progresso amarela
.progressbar--yellow > span {
  background-color: @yellow-color;
}

// Barra de progresso cinza
.progressbar--gray > span {
  background-color: @gray-color;
}

// Barra de progresso branca
.progressbar--white {
  background-color: fade(@white-color, 20%);

  & > span {
    background-color: @white-color;
  }
}

// Barra de progresso sem previsÃ£o para ser resolvida
.progressbar--infinite > span {
  -webkit-animation: progressbar-infinite-animation 1s linear infinite;
  animation: progressbar-infinite-animation 1s linear infinite;
  width: 100% !important;
}





@-webkit-keyframes progressbar-infinite-animation {
  0% {
    -webkit-transform: translate3d(-100%, 0, 0);
    transform: translate3d(-100%, 0, 0)
  }
  100% {
    -webkit-transform: translate3d(100%, 0, 0);
    transform: translate3d(100%, 0, 0)
  }
}

@keyframes progressbar-infinite-animation {
  0% {
    -webkit-transform: translate3d(-100%, 0, 0);
    transform: translate3d(-100%, 0, 0)
  }
  100% {
    -webkit-transform: translate3d(100%, 0, 0);
    transform: translate3d(100%, 0, 0)
  }
}
