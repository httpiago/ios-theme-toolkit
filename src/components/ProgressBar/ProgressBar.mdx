---
name: ProgressBar
route: /components/ProgressBar
menu: Components
---

import { Playground, PropsTable } from 'docz'
import { ProgressBar } from '../../index.js'
import '../../../dist/styles.css'

# ProgressBar

To use this component, simply import it into any file and pass the [modifiers](#/#props) in the properties.

```jsx
import { ProgressBar } from 'ios-theme-toolkit'

export default () => (
  <ProgressBar value={80} />
)
```

## Examples

<Playground>
  <style>{".react-live-preview .progressbar { margin: 24px 0; }"}</style>

  <ProgressBar value={30} />

  <ProgressBar value={80} color="red" />

  <ProgressBar value={50} color="green" height={10} round />

  <ProgressBar value={100} color="yellow" />

  <ProgressBar value={65} color="gray" />

  <ProgressBar color="blue" infinite />
</Playground>

## Props

<PropsTable of={ ProgressBar } />

All the properties which are not listed above will be transferred to the \<div\> tag.
