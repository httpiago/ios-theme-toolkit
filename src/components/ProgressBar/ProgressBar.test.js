/**
 * TEST FILE
 * Please try writing the tests for all component behaviors,
 * manually call the functions within the class, change the properties...
 *
 * @see https://jestjs.io/docs/en/expect
 * @see https://github.com/airbnb/enzyme/tree/master/docs/api
 */
import React from 'react'
import { shallow } from 'enzyme'
import ProgressBar from './ProgressBar'

const COMPONENT = shallow(
  <ProgressBar value={47} className="aditionalClass" />
)

describe('Test ProgressBar component', () => {
  it('should render correctly based on last snapshot', () => {
    expect(COMPONENT).toMatchSnapshot()
  })

  it('should update correctly', () => {
    COMPONENT.setProps({ color: 'red' })
    
    expect(COMPONENT.hasClass('progressbar--red')).toBeTruthy()
    
    // Check if percentage level change correctly
    COMPONENT.setProps({ value: 64 })
    
    expect(COMPONENT.children('span').html()).toContain('width:64%')
  })

  it('should has aditional classes', () => {
    expect(COMPONENT.hasClass('aditionalClass')).toBeTruthy()
  })

  it('should pass aditional props to html element', () => {
    COMPONENT.setProps({ 'data-custom-attr': 'yes' })

    expect(COMPONENT.html()).toContain('data-custom-attr="yes"')
  })

})
