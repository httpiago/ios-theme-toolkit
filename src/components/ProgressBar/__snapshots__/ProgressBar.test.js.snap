// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ProgressBar component should render correctly based on last snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProgressBar
    className="aditionalClass"
    color="blue"
    height={3}
    infinite={false}
    round={false}
    value={47}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <span
        style={
          Object {
            "width": "47%",
          }
        }
      />,
      "className": "progressbar progressbar--blue aditionalClass",
      "style": Object {
        "height": 3,
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "style": Object {
          "width": "47%",
        },
      },
      "ref": null,
      "rendered": null,
      "type": "span",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <span
          style={
            Object {
              "width": "47%",
            }
          }
        />,
        "className": "progressbar progressbar--blue aditionalClass",
        "style": Object {
          "height": 3,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "style": Object {
            "width": "47%",
          },
        },
        "ref": null,
        "rendered": null,
        "type": "span",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
